core/object/message_queue.cpp

```diff
@@ -291,54 +262,203 @@ void MessageQueue::flush() {
+			Message *message = (Message *)&page->data[offset];
+
+			uint32_t advance = sizeof(Message);
+			if ((message->type & FLAG_MASK) != TYPE_NOTIFICATION) {
+				advance += sizeof(Variant) * message->args;
+			}
+
+			Object *target = message->callable.get_object();
+
+			if (target != nullptr) {
+				switch (message->type & FLAG_MASK) {
+					case TYPE_CALL: {
+						if (!call_count.has(message->callable)) {
+							call_count[message->callable] = 0;
+						}
+
+						call_count[message->callable]++;
+
+					} break;
+					case TYPE_NOTIFICATION: {
+						if (!notify_count.has(message->notification)) {
+							notify_count[message->notification] = 0;
+						}
+
+						notify_count[message->notification]++;
+
+					} break;
+					case TYPE_SET: {
+						StringName t = message->callable.get_method();
+						if (!set_count.has(t)) {
+							set_count[t] = 0;
+						}
+
+						set_count[t]++;
+
+					} break;
+				}
+
+			} else {
+				//object was deleted
+				print_line("Object was deleted while awaiting a callback");
+
+				null_count++;
+			}
+
+			//pre-advance so this function is reentrant
+			offset += advance;
+
+			if ((message->type & FLAG_MASK) != TYPE_NOTIFICATION) {
+				Variant *args = (Variant *)(message + 1);
+				for (int k = 0; k < message->args; k++) {
+					args[k].~Variant();
+				}
+			}
+
+			message->~Message();
 		}
 	}
 
+	print_line("TOTAL PAGES: " + itos(pages_used) + " (" + itos(pages_used * QUEUE_PAGE_SIZE) + " bytes).");
+	print_line("NULL count: " + itos(null_count));
+
+	for (const KeyValue<StringName, int> &E : set_count) {
+		print_line("SET " + E.key + ": " + itos(E.value));
+	}
+
+	for (const KeyValue<Callable, int> &E : call_count) {
+		print_line("CALL " + E.key + ": " + itos(E.value));
+	}
+
+	for (const KeyValue<int, int> &E : notify_count) {
+		print_line("NOTIFY " + itos(E.key) + ": " + itos(E.value));
+	}
+}
+
+bool CallQueue::is_flushing() const {
+	return flushing;
+}
+
+int CallQueue::get_max_buffer_usage() const {
+	return pages.size() * QUEUE_PAGE_SIZE;
+}
+
+CallQueue::CallQueue(Allocator *p_custom_allocator, uint32_t p_max_pages, const String &p_error_text) {
+	if (p_custom_allocator) {
+		allocator = p_custom_allocator;
+		allocator_is_custom = true;
+	} else {
+		allocator = memnew(Allocator);
+		allocator_is_custom = false;
+	}
+	max_pages = p_max_pages;
+	error_text = p_error_text;
+}
+
+CallQueue::~CallQueue() {
+	clear();
+	// Let go of pages.
+	for (uint32_t i = 0; i < pages.size(); i++) {
+		allocator->free(pages[i]);
+	}
+	if (!allocator_is_custom) {
+		memdelete(allocator);
+	}
+}
+
+//////////////////////
+
+MessageQueue *MessageQueue::singleton = nullptr;
+
+MessageQueue::MessageQueue() :
+		CallQueue(nullptr, int(GLOBAL_DEF_RST(PropertyInfo(Variant::INT, "memory/limits/message_queue/max_size_mb", PROPERTY_HINT_RANGE, "1,512,1,or_greater"), 32)) * 1024 * 1024 / QUEUE_PAGE_SIZE, "Message queue out of memory. Try increasing 'memory/limits/message_queue/max_size_kb' in project settings.") {
```

Comment:
```suggestion
		CallQueue(nullptr, int(GLOBAL_DEF_RST(PropertyInfo(Variant::INT, "memory/limits/message_queue/max_size_mb", PROPERTY_HINT_RANGE, "1,512,1,or_greater"), 32)) * 1024 * 1024 / QUEUE_PAGE_SIZE, "Message queue out of memory. Try increasing \"memory/limits/message_queue/max_size_mb\" in project settings.") {
```
All this in a constructor call is pretty unreadable though :(